### PR TITLE
mgr: implement completion of osd MetadataUpdate

### PR DESCRIPTION
**Pull request by gmayyyha**
Signed-off-by: Yanhu Cao <gmayyyha@gmail.com>

### CODE REVIEW

File: src/mgr/Mgr.cc [hunk 1]
```diff
@@ -119,7 +121,10 @@ class MetadataUpdate : public Context
         if (daemon_state.exists(key)) {
           state = daemon_state.get(key);
 	  Mutex::Locker l(state->lock);
+          if (key.first == "mds")
           daemon_meta.erase("name");
+          else if (key.first == "osd")
+            daemon_meta.erase("id");
           daemon_meta.erase("hostname");
```
Comment (Liuchang0812) [inline src/mgr/Mgr.cc]: osd's id is integer. we will get a core dump in line139.

File: src/mgr/Mgr.cc [hunk 2]
```diff
@@ -105,6 +105,8 @@ class MetadataUpdate : public Context
                   << key.first << "." << key.second << dendl;
           return;
         }
+        dout(1) << "mon returned valid metadata JSON for "
```
Comment (jcsp) [inline src/mgr/Mgr.cc]: Level 1 is too high for this message, suggest using 4 instead.

File: src/mgr/Mgr.cc [hunk 3]
```diff
@@ -119,7 +121,10 @@ class MetadataUpdate : public Context
         if (daemon_state.exists(key)) {
           state = daemon_state.get(key);
 	  Mutex::Locker l(state->lock);
+          if (key.first == "mds")
           daemon_meta.erase("name");
```
Comment (jcsp) [inline src/mgr/Mgr.cc]: Need to indent this line now that it's inside an `if` (adding some `{}` would also be nice)

**Comment by gmayyyha** [comment]
all done.

**Comment by jcsp** [comment]
@gmayyyha thanks -- I just realised this fixes http://tracker.ceph.com/issues/21159 so could you add a "Fixes: http://tracker.ceph.com/issues/21159" line to the commit message?  That way we can backport it to luminous.

**Comment by gmayyyha** [comment]
@jcsp OK. Done.

**Comment by gmayyyha** [comment]
retest this please.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-08-31_16:26:38-rados-wip-sage-testing-2017-08-31-0909-distro-basic-smithi/

**Pull request merged**